### PR TITLE
[clipboard] Rewrite NCL screen

### PR DESCRIPTION
**Pull request by barthap**
# Why

Resolves ENG-3327

Also, improves the UX of enum buttons - they were difficult to press before if their text is one char long

# How

<details>
<summary>Video 1</summary>

https://user-images.githubusercontent.com/278340/157854864-51ec6299-2bb9-439e-84b7-8ac3a70acb76.mov

</details>

<details>
<summary>Video 2</summary>

https://user-images.githubusercontent.com/278340/157855949-afe2e60f-fb9a-482a-86e1-89aafe3f98f1.mp4

</details>

# Test Plan

> CI might fail, because this branch doesn't include required changes from #16581 - will need to rebase it later, but I want to merge #16551 first

### CODE REVIEW

File: apps/native-component-list/src/components/FunctionDemo/EnumButton.tsx [hunk 1]
```diff
@@ -59,6 +59,8 @@ const styles = StyleSheet.create({
     paddingHorizontal: 6,
     backgroundColor: Colors.tintColor,
     borderRadius: 5,
+    minWidth: 30,
+    alignItems: 'center',
```
Comment (barthap) [inline apps/native-component-list/src/components/FunctionDemo/EnumButton.tsx:63]: Changes in this file make enum buttons a bit wider and easier to click when the button text is only one character long (`0`, `1` etc.)

File: apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        },
+      ],
+    },
+  ],
+  actions: Clipboard.getImageAsync,
+};
+
+const SET_URL_ASYNC_CONFIG: FunctionDescription = {
+  name: 'setUrlAsync',
+  platforms: ['android'],
```
Comment (barthap) [inline apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx:154]:
```suggestion
  platforms: ['ios'],
```
leftover from experiments

File: apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+import ClipboardListenerDemo from './ClipboardListenerDemo';
+import ImagePreview from './ImagePreview';
+
+const withSupportedPlatforms = (platforms: Platform[], action: ActionFunction): ActionFunction =>
+  isCurrentPlatformSupported(platforms)
+    ? action
+    : () => Promise.resolve('Function is not supported on this platform');
+
+const HAS_X_ASYNC_CONFIG: FunctionDescription = {
+  name: 'hasXAsync',
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx:18]: I'm not sure about this `X`, it's a little confusing, but I don't really have any better idea 👅 
`has<type>Async`, `has{String, Image, Url}Async`, `has{X}Async`? 🤔

File: apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+      const result = await ImagePicker.launchImageLibraryAsync({
+        mediaTypes: ImagePicker.MediaTypeOptions.Images,
+        base64: true,
+      });
+      if (!result.cancelled && result.base64) {
+        await Clipboard.setImageAsync(result.base64);
+        return 'Image copied to clipboard';
+      }
+    } else {
+      Alert.alert('Permission required!', 'You must allow accessing images in order to proceed.');
```
Comment (bbarthec) [inline apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx:115]: I dislike this one 😞 
I would not use `Alert` or any other popup 🤔 
Let's go with simple rejection maybe for now?
We need to add mechanism for handling failing cases, like missing prerequisites or promise rejections. Ideally we could log more info in the console, like what's the actual method signature that failed, etc 😁 
When we add such mechanism these rejected function calls would be presented gracefully as well 💪

**Review by bbarthec — COMMENTED** [review]
Looking good 👍 
My remarks/questions:
- as `hasXAsync` is either available or not on the current platform depending on the function used under the hood I would rather split them to the separate parts. I dislike that you propose a different way to communicate platform unavailability (resolving promise with some message). What would you say?
- what do you about `FunctionDemo` mechanism? Is it usable or should we refactor it somehow?

File: apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx (hunk 3 above)
Comment (barthap) [reply apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx:18]: `hasSomethingAsync`?

File: apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx (hunk 4 above)
Comment (barthap) [reply apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx:115]: For now, I'd just return this message to be displayed in the result box

**Review by barthap — COMMENTED** [review]
> as hasXAsync is either available or not on the current platform depending on the function used under the hood I would rather split them to the separate parts. I dislike that you propose a different way to communicate platform unavailability (resolving promise with some message). What would you say?

I was afraid that you'll say this 🙄 

I was really happy that with multiple `actions` I could just put all three very similar functions into one place instead of making separate sections for each one of them - I saved so much space and the whole screen looks much more concise.

The solution would be to implement platform-specific actions in FunctionDemo. That shouldn't be hard, but it's out of this PR's scope.

> what do you about FunctionDemo mechanism? Is it usable or should we refactor it somehow?

It turns out to be really handy 🎉 and its components can easily be reused in something custom (see Clipboard Listener). The only thing I'm missing is maybe a _foldable section_ component to organize functions into groups - e.g. here: strings, images, urls

File: apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx (hunk 4 above)
Comment (tsapeta) [reply apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx:115]: To be honest I don't really get that why you're against all alerts/popups. I agree that in many cases they would be annoying (especially for correct and expected results), but here this is more like a rare exception and in my opinion presenting an alert is much better UX than looking for console logs.

File: apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+      action: withSupportedPlatforms(['android', 'ios'], Clipboard.hasImageAsync),
+    },
+    {
+      name: 'hasUrlAsync',
+      action: withSupportedPlatforms(['ios'], Clipboard.hasImageAsync),
+    },
+  ],
+};
+
+const SET_STRIING_ASYNC_CONFIG: FunctionDescription = {
```
Comment (tsapeta) [inline apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx:36]:
```suggestion
const SET_STRING_ASYNC_CONFIG: FunctionDescription = {
```

File: apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx (hunk 3 above)
Comment (tsapeta) [reply apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx:18]: What if we split this config into three parts, separate for each method?

File: apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+const GET_URL_ASYNC_CONFIG: FunctionDescription = {
+  name: 'getUrlAsync',
+  platforms: ['ios'],
+  parameters: [],
+  actions: Clipboard.getUrlAsync,
+};
+
+const FUNCTIONS_DESCRIPTIONS = [
+  HAS_X_ASYNC_CONFIG,
+  SET_STRIING_ASYNC_CONFIG,
```
Comment (tsapeta) [inline apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx:174]:
```suggestion
  SET_STRING_ASYNC_CONFIG,
```

File: apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx (hunk 3 above)
Comment (barthap) [reply apps/native-component-list/src/screens/Clipboard/ClipboardScreen.tsx:18]: > What if we split this config into three parts, separate for each method?

https://github.com/expo/expo/pull/16594#pullrequestreview-907167577

**Review by tsapeta — APPROVED** [review]

**Pull request merged**